ClangSharpPInvokeGenerator/CXTypeExtensions.cs

```diff
@@ -0,0 +1,408 @@
+
+            switch (callingConvention)
+            {
+                case CXCallingConv.CXCallingConv_C:
+                {
+                    return "Cdecl";
+                }
+
+                default:
+                {
+                    Debug.WriteLine($"Unhandled calling convention: {callingConvention} in {type.KindSpelling} in {cursor.KindSpelling}.");
+                    Debugger.Break();
+                    return string.Empty;
+                }
+            }
+        }
+
+        public static string GetMarshalAttribute(this CXType type, CXCursor cursor)
+        {
+            switch (type.kind)
+            {
+                case CXTypeKind.CXType_Void:
+                case CXTypeKind.CXType_UShort:
+                case CXTypeKind.CXType_UInt:
+                case CXTypeKind.CXType_ULong:
+                case CXTypeKind.CXType_ULongLong:
+                case CXTypeKind.CXType_Short:
+                case CXTypeKind.CXType_Int:
+                case CXTypeKind.CXType_Long:
+                case CXTypeKind.CXType_LongLong:
+                case CXTypeKind.CXType_Double:
+                case CXTypeKind.CXType_Enum:
+                case CXTypeKind.CXType_Typedef:
+                case CXTypeKind.CXType_ConstantArray:
+                {
+                    return string.Empty;
+                }
+
+                case CXTypeKind.CXType_Pointer:
+                {
+                    return type.GetMarshalAttributeForPointeeType(cursor, type.PointeeType);
+                }
+
+                case CXTypeKind.CXType_Elaborated:
+                {
+                    return type.CanonicalType.GetMarshalAttribute(cursor);
+                }
+
+                default:
+                {
+                    Unhandled(type, cursor);
+                    return string.Empty;
+                }
+            }
+        }
+
+        public static string GetMarshalAttributeForPointeeType(this CXType type, CXCursor cursor, CXType pointeeType)
+        {
+            switch (pointeeType.kind)
+            {
+                case CXTypeKind.CXType_Void:
+                case CXTypeKind.CXType_UShort:
+                case CXTypeKind.CXType_UInt:
+                case CXTypeKind.CXType_ULong:
+                case CXTypeKind.CXType_ULongLong:
+                case CXTypeKind.CXType_Short:
+                case CXTypeKind.CXType_Int:
+                case CXTypeKind.CXType_Long:
+                case CXTypeKind.CXType_LongLong:
+                case CXTypeKind.CXType_Double:
+                case CXTypeKind.CXType_Pointer:
+                case CXTypeKind.CXType_Enum:
+                case CXTypeKind.CXType_Typedef:
+                case CXTypeKind.CXType_FunctionProto:
+                {
+                    return string.Empty;
+                }
+
+                case CXTypeKind.CXType_Char_S:
+                {
+                    return "MarshalAs(UnmanagedType.CustomMarshaler, MarshalTypeRef = typeof(StringMarshaler))";
```

Comment:
This could be `LPUTF8Str` if we said `net48`, `netstandard2.1`, and `netcoreapp10` are the baseline